bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -2741,19 +2740,24 @@ private static boolean isEqual(TableValueImpl lhsTable, TableValueImpl rhsTable,
             return false;
         }
 
-        if (((BTableType) lhsTable.getType()).getFieldNames() != null &&
-                ((BTableType) lhsTable.getType()).getFieldNames().length > 0) {
-            for (Map.Entry<BAnydataType, Object> lhsTableEntry :
-                    (Iterable<Map.Entry<BAnydataType, Object>>) lhsTable.entrySet()) {
-                if (!isEqual(lhsTableEntry.getValue(), rhsTable.get(lhsTableEntry.getKey()), checkedValues)) {
+        boolean isLhsKeyedTable = ((BTableType) lhsTable.getType()).getFieldNames() != null &&
+                ((BTableType) lhsTable.getType()).getFieldNames().length > 0;
+        boolean isRhsKeyedTable = ((BTableType) rhsTable.getType()).getFieldNames() != null &&
+                ((BTableType) rhsTable.getType()).getFieldNames().length > 0;
+
+        Object[] lhsTableValues = lhsTable.values().toArray();
+        Object[] rhsTableValues = rhsTable.values().toArray();
+
+        if (isLhsKeyedTable == isRhsKeyedTable) {
+            for (int i = 0; i < lhsTableValues.length; i++) {
```

Comment:
If the table sizes are unequal, can return fast

Reply:
Returns early at https://github.com/ballerina-platform/ballerina-lang/blob/master/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java#L2740